docs/RFCS/20221124_split_tls.md

```diff
@@ -0,0 +1,295 @@
+  fields during the request), making it difficult to issue and utilize
+  those certificates
+
+In the past, we already have acknowledged these problems in the
+context of HTTP clients. Because we took these concerns seriously, we
+allowed users to provision a fully separate TLS server cert (with its
+own CA) for the HTTP service. However, more recently we have learned
+that the same concern exists also for the SQL service and the RPC
+service as used by CLI clients (cockroach node decommission, etc).
+
+Moving away from using the certificates used for node-to-node
+communication allows database operators to use their existing PKI (or
+automation such as LetsEncrypt) to manage the certificates their
+clients connect to, and simplifies future work to improve how the
+certificates used for node-to-node communication can be generated.
+
+# Technical design
+
+Currently the cert manager code inside CockroachDB has just two APIs
+for initializing+reading a 'server cert':
+
+- `GetUITLSconfig`, used by the HTTP service.
+- GetTLSConfig, used by all of:
+  - node-node RPC service
+  - node-client RPC service and RPC client
+  - node-client SQL service
+
+We will change this as follows:
+- `GetUITLSconfig` - unchanged
+- `GetTLSConfig`  gets split into:
+  - `GetRPCServerTLSConfig`, used for both node-node and node-client RPC service
+     Bundles 2 server peer certs in the TLSConfig, as specified below.
+  - `GetSQLServerTLSConfig`, used for the SQL service.
+  - `GetRPCClientTLSConfig`, used for RPC clients (client side of node-node connections, and also CLI clients)
+
+![server certs](20221124_split_tls/server_certs.png)
+
+If the `ui.crt` doesn’t exist, the HTTPS endpoint will fallback to the
+new `server.crt`, allowing database operators who don’t need a
+separation of those interfaces to use the same certificate for all
+three.
+
+The CA story is complementary to the story above: the `tls.Config`
+needs to know which CA cert to use to validate the _peer_ (the other
+side of the connection / the client). For each of the 3 services, and
+the RPC client, we bundle CAs as follows:
+
+![CA certs](20221124_split_tls/ca_certs.png)
+
+The diagrams above represent the desired end state of the
+functionality. Note that the code used to generate the `tls.Config` for
+SQL clients has been omitted: it is owned by the SQL client driver
+(e.g. pgx, lib/pq) and we don't control it.
+
+## Changes from current implementation
+
+To enumerate the actual changes, we also need to look at the current implementation.
+
+![server certs - orig](20221124_split_tls/server_certs_orig.png)
+
+![CA certs - orig](20221124_split_tls/ca_certs_orig.png)
+
+
+Enumeration of changes:
+
+- `GetTLSConfig` becomes split into:
+  - `GetSQLServerTLSConfig`
+  - `GetRPCServerTLSConfig`
+  - `GetRPCClientTLSConfig`
+
+- Each defined using the lookup logic defined above.
+
+- In particular:
+  - `ca-client.crt` becomes mandatory. There's a one-time copy of
```

Comment:
Why are we making split client & server CA mandatory? What does that have to do with the motivation to separate node and server certs? At this point, we warn customers to do that only when really necessary as per their needs - https://www.cockroachlabs.com/docs/stable/authentication.html#using-split-ca-certificates.

Reply:
Existing organizations and organizations that have effectively managed PKI will be able to use the same certificate authority for both, by using the same CA for both files. This is more for new organizations, especially those that are new to certificate-based authentication (and new to running PKI that issues user client certificates). If the same certificate authority issues both node-to-node certificates and client certificates, even if there are are intermediate certificates involved, preventing misuse of certificates intended for users vs certificates intended for machines will be very difficult. We cannot require organizations add custom fields to their issued certificates to help our code determine whether or not a certificate is intended for user authentication or node-to-node authentication, so this is the next best solution.

Reply:
If the cluster isn't doing certificate backed authentication then I'd argue we shouldn't require `ca-client.crt`. 

We currently manage all of the certificates involved in CC, so this would have no impact on what we're doing, aside from opening up the future possibility of customers uploading a trusted CA certificate if they didn't want to use ours.

Reply:
@knz pointed out that we use client certificates as a recovery method, so we can't make them optional unfortunately, but I'd argue between the copying/linking of the original `ca.crt` to `ca-client.crt` for clusters without an existing `ca-client.crt` and the `cockroach cert` tool for organizations that don't have their own PKI setup I don't believe requiring this file will be a huge issue

Reply:
Well my concern is primarily from a usability perspective. Plus the aspect of recovery method applies today too but we still don't mandate the client CA cert. IMO, it is worse than making something optional (like in case of ca.crt). 

At the same time I don't know the deployment profile of majority of our self-hosted customers' well enough to decide whether I should be strongly opinionated about it or not. Hence I'm considering setting up a meeting with a few CEAs to review the salient aspects of this RFC along with you and @knz.